extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/common/K8sTaskAdapter.java

```diff
@@ -0,0 +1,277 @@
+    String contents = taskJson.map(envVar -> taskJson.get().getValue()).orElse(null);
+    if (contents == null) {
+      throw new IOException("No TASK_JSON environment variable found in pod: " + from.getMetadata().getName());
+    }
+    return mapper.readValue(Base64Compression.decompressBase64(contents), Task.class);
+  }
+
+  @VisibleForTesting
+  public abstract Job createJobFromPodSpec(PodSpec podSpec, Task task, PeonCommandContext context) throws IOException;
+
+  protected Job buildJob(
+      K8sTaskId k8sTaskId,
+      Map<String, String> labels,
+      Map<String, String> annotations, PodTemplateSpec podTemplate
+  )
+  {
+    return new JobBuilder()
+        .withNewMetadata()
+        .withName(k8sTaskId.getK8sTaskId())
+        .addToLabels(labels)
+        .addToAnnotations(annotations)
+        .endMetadata()
+        .withNewSpec()
+        .withTemplate(podTemplate)
+        .withActiveDeadlineSeconds(config.maxTaskDuration.toStandardDuration().getStandardSeconds())
+        .withBackoffLimit(0)
+        .withTtlSecondsAfterFinished((int) config.taskCleanupDelay.toStandardDuration().getStandardSeconds())
+        .endSpec()
+        .build();
+  }
+
+  @VisibleForTesting
+  static Optional<Long> getJavaOptValueBytes(String qualifier, List<String> commands)
+  {
+    Long result = null;
+    Optional<String> lastHeapValue = commands.stream().filter(x -> x.startsWith(qualifier)).reduce((x, y) -> y);
+    if (lastHeapValue.isPresent()) {
+      result = HumanReadableBytes.parse(StringUtils.removeStart(lastHeapValue.get(), qualifier));
+    }
+    return Optional.ofNullable(result);
+  }
+
+  @VisibleForTesting
+  static long getContainerMemory(PeonCommandContext context)
+  {
+    List<String> javaOpts = context.getJavaOpts();
+    Optional<Long> optionalXmx = getJavaOptValueBytes("-Xmx", javaOpts);
+    long heapSize = HumanReadableBytes.parse("1g");
+    if (optionalXmx.isPresent()) {
+      heapSize = optionalXmx.get();
+    }
+    Optional<Long> optionalDbb = getJavaOptValueBytes("-XX:MaxDirectMemorySize=", javaOpts);
+    long dbbSize = heapSize;
+    if (optionalDbb.isPresent()) {
+      dbbSize = optionalDbb.get();
+    }
+    return (long) ((dbbSize + heapSize) * 1.2);
+
+  }
+
+  protected void setupPorts(Container mainContainer)
+  {
+    mainContainer.getPorts().clear();
+    ContainerPort tcpPort = new ContainerPort();
+    tcpPort.setContainerPort(DruidK8sConstants.PORT);
+    tcpPort.setName("druid-port");
+    tcpPort.setProtocol("TCP");
+    ContainerPort httpsPort = new ContainerPort();
+    httpsPort.setContainerPort(DruidK8sConstants.TLS_PORT);
+    httpsPort.setName("druid-tls-port");
+    httpsPort.setProtocol("TCP");
+    mainContainer.setPorts(Lists.newArrayList(httpsPort, tcpPort));
+  }
+
+  protected void addEnvironmentVariables(Container mainContainer, PeonCommandContext context, String taskContents)
```

Comment:
similar to the other comments, I think we want to make additional things configurable here if possible

Reply:
agreed, i will make env's and annotations configurable.

Reply:
actually one quesiton, for env variables being configurable, do you have a usecase?  It takes all the env variables of the overlord pod running, would there be a situation where you would need something additional here?  The container shouldn't need more than what druid has to work, right?

Reply:
annotations and labels are configurable, if you can provide me with an example of how we would need env variables, i would be happy to add it as configurable.  But I personally think that could cause more harm than good potentially.

Reply:
I thought I added it as a comment but what I actually think is most important to be configurable is nodeSelector/affinity/tolerations in case people want to configure what node pools to run on.

If nodeSelector is not set than maybe we can use whatever the overlord uses as its node selector.

Env variables is not as important, but I could see something like DD_TRACE_AGENT_URL for enabling tracing with datadog

Reply:
you have a valid point, maybe in a future PR, we could add some config map where you can define a podSpec template.  That should be pretty easy to do.

Reply:
may not be an issue but is there a limit in k8s on how large the env variable value can be? The task JSON could be large.

Reply:
There is and its no different from a configmap max size either.  That is why I compress and base64 encode the task.json file when i pass it in the spec.